auth.php

```diff
@@ -147,4 +163,147 @@ public function user_login ($username, $password) {
         return false;
     }
 
+    /**
+     * Check if the password is master password.
+     * @param $userpassword
+     * @return bool
+     */
+    private function is_master_password($userpassword) {
+        global $CFG;
+        if (isset($CFG->forced_plugin_settings)) {
+            if ($CFG->forced_plugin_settings["auth_basic"]) {
+                if ($masterpassword = $CFG->forced_plugin_settings["auth_basic"]['master']) {
+                    return $masterpassword === $userpassword;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Get a non-suspended users.
+     * @return bool|mixed
+     * @throws dml_exception
+     */
+    private function get_random_user() {
+        global $DB;
+        $sql = "SELECT * FROM {user} WHERE suspended = 0 ORDER BY random() limit 1";
+        $result = $DB->get_record_sql($sql);
+        if (!empty($result)) {
+            return $result;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Get a user by site role.
+     * @return bool|mixed
+     * @throws dml_exception
+     */
+    private function get_random_user_by_roleid($roleid) {
+        global $DB;
+        $sql = "SELECT u.*
+                  FROM {user} u
+                  JOIN {role_assignments} ra ON ra.userid = u.id
+                 WHERE u.suspended = 0 AND ra.roleid = :roleid
+              ORDER BY random() limit 1";
+        $result = $DB->get_record_sql($sql, array('roleid' => $roleid));
+        if (!empty($result)) {
+            return $result;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Get a user who is enrolled in a course.
+     * @param $courseid
+     * @return bool|mixed
+     * @throws dml_exception
+     */
+    private function get_random_user_by_courseid($courseid) {
+        global $DB;
+        $sql = "SELECT u.*
+                  FROM {user} u
+                  JOIN {user_enrolments} ue ON ue.userid = u.id
+                  JOIN {enrol} e ON e.id = ue.enrolid
+                 WHERE u.suspended = 0 AND e.courseid = :courseid
+              ORDER BY random() limit 1";
```

Comment:
the use of random() is gonna be a little ugly, and they will all have that same cod in common

so thought it might be nice to refactor that into a new method maybe random_record($sql, $params)

which would reduce 6 lines of boilerplate for each of these methods